### PR TITLE
=htc #416 fix for connection pool crash

### PR DESCRIPTION
**Pull request by gosubpl**
This is to fix connection pool crashing entirely in case of response entity future failing.

### CODE REVIEW

**Comment by gosubpl** [comment]
Refs #416

**Comment by gosubpl** [comment]
**Please do not merge before reading this note**
As per the discussion on the issue ticket (https://github.com/akka/akka-http/issues/416) I have implemented only a simple fix that results in that in most cases the state will end up in `Idle` instead of more proper `Disconnect`. While functionally this is not a problem, the connection slot scheduler prefers `Idle` slots to `Disconnected` ones as they are supposedly hot.
I have submitted this PR to not block any progress on `FIRST_STABLE` release.

- If the release train needs to depart, this PR can be merged and I will register an issue to fix the terminal state of the connection pool, as suggested by @leachbj (that can be solved with doing the below)
- If the release will not happen anytime soon, please let me know and I will extend connection pool state machine logic with additional processing based on idea inspired by `SlotEvent.NoOp` (https://github.com/leachbj/akka-http/commit/516f997b94fb980a1a4b314ed3bd4c5b49a9597d) to get proper handling for the terminal state of the pool in this case.

@jrudolph @leachbj many thanks for the insightful discussion leading to this PR. Also @jrudolph many thanks for your reproducer with tracing - investigating that issue with this was a breeze (relatively speaking :) )!

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala [hunk 1]
```diff
@@ -150,28 +150,28 @@ private object PoolConductor {
...
 
         // eagerly start at least slotSettings.minSlots connections
         (0 until slotSettings.minSlots).foreach { connect }
       }
 
       def connect(slotIx: Int): Unit = {
-        emit(out, SwitchSlotCommand(ConnectEagerlyCommand, slotIx))
+        emit(slotCommandsOut, SwitchSlotCommand(ConnectEagerlyCommand, slotIx))
         slotStates(slotIx) = Idle
       }
```
Comment (leachbj) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala]: I would recommend leaving these changes out as they don't contribute to the fix (they're good changes but unrelated to the fix).

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala [hunk 2]
```diff
@@ -112,7 +113,24 @@ private object PoolSlot {
...
+          //    SlotEvent.RequestCompleted(slot); transitions from Loaded(1) => Idle
+          // 2. Fail after the response headers (Future failed and then onUpstreamFailure);
+          //    SlotEvent.RequestCompleted(slot); transitions from Loaded(1) => Idle
+          //    SlotEvent.Disconnected(slot, 0); transitions from Idle => Unconnected
+          // the result is that the final state of the slot is usually Idle which results in this slot interpreted as hot
+          // and being preferred by the scheduler instead of more correct Unconnected
+          // FIXME: enhance the event & processing in the connection pool state machine to allow proper transition
+          val completed = whenCompleted.map(_ ⇒ SlotEvent.RequestCompleted(slotIx))
+            .recoverWith { case _ ⇒ FastFuture.successful(SlotEvent.RequestCompleted(slotIx)) }
+          push(eventsOut, SlotEvent.RequestCompletedFuture(completed))
```
Comment (leachbj) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala]: 👍

**Comment by ktoso** [comment]
Disabled the troublesome node (a7).

**Comment by ktoso** [comment]
PLS BUILD

**Comment by gosubpl** [comment]
@leachbj many thanks for reviewing. I have moved the `PoolConductor` logging improvements to a separate ticket (https://github.com/akka/akka-http/issues/482 ).

**Comment by jrudolph** [comment]
I think the fix is more important than perfect sync as long as we keep in mind that we might want to improve the situation (or that we might have missed cases). I created #490 to track this and also added the link to the FIXME. No waiting for PR validation and then I'll merge it.

**Comment by jrudolph** [comment]
Thanks, @gosubpl and @leachbj for great discussion and help with this one.

**Pull request merged**